base/server/python/pki/server/deployment/scriptlets/keygen.py

```diff
@@ -93,23 +93,44 @@ def generate_csr(self,
         (key_type, key_size, curve, hash_alg) = self.get_key_params(
             deployer, cert_id)
 
-        nssdb.create_request(
-            subject_dn=subject_dn,
-            request_file=csr_path,
-            key_type=key_type,
-            key_size=key_size,
-            curve=curve,
-            hash_alg=hash_alg,
-            basic_constraints_ext=basic_constraints_ext,
-            key_usage_ext=key_usage_ext,
-            extended_key_usage_ext=extended_key_usage_ext,
-            subject_key_id=subject_key_id,
-            generic_exts=generic_exts)
-
-        with open(csr_path) as f:
-            csr = f.read()
-
-        b64_csr = pki.nssdb.convert_csr(csr, 'pem', 'base64')
+        """
+
+cfu::: for KRA, storage cert and transport cert need to use the new -w
+option of PKCS10Client
+e.g. PKCS10Client -d /var/lib/pki/<ca instance>/alias -h hsm-module -a rsa -l 2048 -n "CN=CA issuanceProt cert" -w -v -o ca-issuanceProt-cfu.csr.b64
+
+        """
+
+        print('pki_hsm_enable: ', config.str2bool(deployer.mdict['pki_hsm_enable']))
+        print('subsystem.type: ', subsystem.type)
+        if (subsystem.type == 'KRA' and 
+            config.str2bool(deployer.mdict['pki_hsm_enable']) and
+                (cert_id in ['storage', 'transport'])):
```

Comment:
Should it check to make sure that `key_type` is `rsa`? If it's not `rsa`, should it proceed to create an RSA key anyway or should it fail?